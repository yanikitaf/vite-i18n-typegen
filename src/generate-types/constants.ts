export const MESSAGES = {
  ERROR: {
    NO_INPUT_DIR:
      '❌ Укажите путь к директории через аргумент или в конфигурационном файле!',
    INVALID_DIR: '❌ Указанная директория не существует или недоступна!',
    GENERIC: '❌ Ошибка при генерации типов:',
  },
  SUCCESS: '✅ Файл translations.d.ts успешно сгенерирован в',
};
