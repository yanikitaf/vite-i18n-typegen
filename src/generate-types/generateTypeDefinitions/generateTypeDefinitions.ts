import type { Config } from '../../types.js';

type TranslationDescriptions = Record<string, Record<string, string>>;

const createParamType = (params: string[]): string => {
  const uniqueParams = [...new Set(params)];
  return uniqueParams.length > 0
    ? `{ ${uniqueParams.map((p) => `${p}: string | number | Date`).join('; ')} }`
    : 'undefined';
};

const generateJsDoc = (
  key: string,
  langs: Record<string, string>,
  preferredLangOrder: string[],
): string => {
  if (!langs || Object.keys(langs).length === 0) {
    return `  /** No translations found for key: ${key} */`;
  }

  const allEntries = [
    ...preferredLangOrder
      .filter((lang) => langs[lang])
      .map((lang) => [lang, langs[lang]] as [string, string]),
    ...Object.entries(langs).filter(
      ([lang]) => !preferredLangOrder.includes(lang),
    ),
  ];

  const escapeBraces = (text: string) =>
    text.replace(/{/g, '{{').replace(/}/g, '}}');

  const previewLine = allEntries.map(([_, v]) => escapeBraces(v)).join(' / ');

  const columnSizes = allEntries.reduce(
    (acc, [lang, text]) => ({
      lang: Math.max(acc.lang, lang.length),
      text: Math.max(acc.text, escapeBraces(text).length),
    }),
    { lang: 4, text: 11 },
  );

  const header = `| ${'lang'.padEnd(columnSizes.lang)} | ${'translation'.padEnd(columnSizes.text)} |`;
  const divider = `| ${'-'.repeat(columnSizes.lang)} | ${'-'.repeat(columnSizes.text)} |`;
  const rows = allEntries.map(
    ([lang, text]) =>
      `| ${lang.padEnd(columnSizes.lang)} | ${escapeBraces(text).padEnd(columnSizes.text)} |`,
  );

  return [
    `  /**`,
    `   * ${previewLine}`,
    `   *`,
    `   * @translations`,
    `   * ${header}`,
    `   * ${divider}`,
    ...rows.map((line) => `   * ${line}`),
    `   */`,
  ].join('\n');
};

const generateTranslationKeys = (
  keyParamsMap: Record<string, string[]>,
): string => {
  const keys = Object.keys(keyParamsMap);

  if (keys.length === 0) {
    return '';
  }

  if (keys.length === 1) {
    return `export type TranslationKeys = "${keys[0]}"`;
  }

  const keyStrings = keys.map((key) => `| "${key}"`);
  const keyString = keyStrings.join('\n');

  return `export type TranslationKeys = \n${keyString};`;
};

const generateTranslationParamsMap = (
  keyParamsMap: Record<string, string[]>,
): string => {
  const paramDefinitions = Object.entries(keyParamsMap)
    .map(([key, params]) => {
      const paramType = createParamType(params);
      return `  "${key}": ${paramType};`;
    })
    .join('\n');

  if (!paramDefinitions) {
    return '';
  }

  return `export type TranslationParamsMap = {\n${paramDefinitions}\n};`;
};

const generateTranslateFunctionDocs = (
  keyParamsMap: Record<string, string[]>,
  descriptions: TranslationDescriptions,
  preferredLangOrder: string[],
): string => {
  const functionDocs = Object.entries(keyParamsMap)
    .map(([key, params]) => {
      const hasParams = params.length > 0;
      const paramType = hasParams ? createParamType(params) : 'undefined';
      const jsDoc = descriptions[key]
        ? generateJsDoc(key, descriptions[key], preferredLangOrder)
        : `/** Missing translations for key: ${key} */`;
      return `${jsDoc}\n(key: "${key}"${hasParams ? `, params: ${paramType}` : ''}): string;`;
    })
    .join('\n\r');

  return `export interface TranslateFunctionDocs {\n${functionDocs}\n}`;
};

export const generateTypeDefinitions = (
  keyParamsMap: Record<string, string[]>,
  descriptions: TranslationDescriptions,
  config: Config,
): string => {
  if (Object.keys(keyParamsMap).length === 0) {
    throw new Error(`No translation keys found}`);
  }

  const typeDefinitions = [];

  if (config.generateKeys) {
    typeDefinitions.push(generateTranslationKeys(keyParamsMap));
  }

  if (config.generateParams) {
    typeDefinitions.push(generateTranslationParamsMap(keyParamsMap));
  }

  if (config.generateDocs) {
    typeDefinitions.push(
      generateTranslateFunctionDocs(
        keyParamsMap,
        descriptions,
        config.preferredLangOrder ?? [],
      ),
    );
  }

  return typeDefinitions.join('\n\r');
};
