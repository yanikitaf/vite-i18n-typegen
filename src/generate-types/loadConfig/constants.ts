import type { Config } from '../../types.js';

export const DEFAULT_CONFIG: Config = {
  inputDir: null,
  outputPath: './',
  outputFileName: 'translations.d.ts',
  jsonFileExtension: '.json',
};
