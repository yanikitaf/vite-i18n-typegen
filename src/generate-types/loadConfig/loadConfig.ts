import { promises as fs } from 'fs';
import path from 'path';
import type { Config } from '../../types.js';
import { DEFAULT_CONFIG } from './constants.js';

export const loadConfig = async (): Promise<Config> => {
  console.log('Загрузка конфигурации...');
  const configPath = path.join(process.cwd(), 'translation-config.json');
  try {
    await fs.access(configPath);
    const content = await fs.readFile(configPath, 'utf-8');
    return { ...DEFAULT_CONFIG, ...JSON.parse(content) };
  } catch {
    return DEFAULT_CONFIG;
  }
};
