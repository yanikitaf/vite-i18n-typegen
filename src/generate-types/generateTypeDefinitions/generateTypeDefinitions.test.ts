import { describe, test, expect } from 'vitest';
import { generateTypeDefinitions } from './generateTypeDefinitions';
import { DEFAULT_CONFIG } from '../loadConfig/constants';

describe('generateTypeDefinitions', () => {
  test('Генерирует корректные типы и документацию', () => {
    const keyParamsMap = {
      'common.welcome': [],
      'products.cart.items': ['count'],
      'mixed.params': ['id', 'total'],
    };

    const descriptions = {
      'common.welcome': {
        ru: 'Добро пожаловать!',
        en: 'Welcome!',
      },
      'products.cart.items': {
        ru: '{{count}} товаров',
        en: '{{count}} items',
      },
      'mixed.params': {
        en: 'ID: {{id}}, Total: {{total}}',
        de: 'ID: {{id}}',
      },
    };

    const result = generateTypeDefinitions(
      keyParamsMap,
      descriptions,
      DEFAULT_CONFIG,
    );

    // Check if all expected keys are present in the TranslationKeys type
    expect(result).toContain('export type TranslationKeys =');
    expect(result).toContain('"common.welcome"');
    expect(result).toContain('"products.cart.items"');
    expect(result).toContain('"mixed.params"');

    // Check if the generated params are correct
    expect(result).toContain('count: string | number | Date');
    expect(result).toContain('id: string | number | Date');
    expect(result).toContain('total: string | number | Date');
  });

  test('Соблюдает порядок языков согласно preferredLangOrder', () => {
    const keyParamsMap = { 'test.key': [] };
    const descriptions = {
      'test.key': {
        de: 'German',
        fr: 'French',
        ru: 'Russian',
        en: 'English',
      },
    };

    const result = generateTypeDefinitions(keyParamsMap, descriptions, {
      ...DEFAULT_CONFIG,
      preferredLangOrder: ['fr', 'de'],
    });

    const expectedOrder = [
      'fr     | French',
      'de     | German',
      'en     | English',
      'ru     | Russian',
    ].join('\n');

    expect(result).toMatch(new RegExp(expectedOrder));
  });

  test('Объединяет параметры из всех языков', () => {
    const keyParamsMap = { 'test.key': ['count'] };
    const descriptions = {
      'test.key': {
        en: '{count} {total}',
        ru: '{count}',
      },
    };

    const result = generateTypeDefinitions(
      keyParamsMap,
      descriptions,
      DEFAULT_CONFIG,
    );
    expect(result).toMatchSnapshot();
    expect(result).toMatch(/{ count: string \| number \| Date }/);
    // expect(result).not.toMatch(/total/);
  });

  test('Выбрасывает ошибку при отсутствии ключей', () => {
    expect(() => generateTypeDefinitions({}, {}, DEFAULT_CONFIG)).toThrow(
      'No translation keys found',
    );
  });

  test('Корректно обрабатывает отсутствующие описания', () => {
    const keyParamsMap = { 'missing.key': ['param'] };
    const result = generateTypeDefinitions(keyParamsMap, {}, DEFAULT_CONFIG);

    expect(result).toContain(
      '/** Missing translations for key: missing.key */',
    );
    expect(result).toMatch(/params: { param: string \| number \| Date }/);
  });

  test('Форматирует таблицу с разными длинами текста', () => {
    const keyParamsMap = { 'long.key': [] };
    const descriptions = {
      'long.key': {
        en: 'Very long translation text',
        ru: 'Очень длинный текст перевода',
      },
    };

    const result = generateTypeDefinitions(
      keyParamsMap,
      descriptions,
      DEFAULT_CONFIG,
    );
    expect(result).toContain('| en   | Very long translation text   |');
    expect(result).toContain('| ru   | Очень длинный текст перевода |');
  });

  test('Snapshot: полная структура', () => {
    const keyParamsMap = {
      'common.welcome': [],
    };

    const descriptions = {
      'common.welcome': {
        en: 'Welcome!',
        ru: 'Добро пожаловать!',
      },
      'products.cart.items': {
        en: '{count} items',
        ru: '{count} товаров',
      },
    };

    const result = generateTypeDefinitions(
      keyParamsMap,
      descriptions,
      DEFAULT_CONFIG,
    );
    expect(result).toMatchSnapshot();
  });
});
