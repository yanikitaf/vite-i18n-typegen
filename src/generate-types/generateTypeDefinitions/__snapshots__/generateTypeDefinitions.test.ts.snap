// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateTypeDefinitions > Snapshot: полная структура 1`] = `
"export type TranslationKeys = "common.welcome"

export type TranslationParamsMap = {
  "common.welcome": undefined;
};

export interface TranslateFunctionDocs {
  /**
   * Welcome! / Добро пожаловать!
   *
   * @translations
   * | lang | translation       |
   * | ---- | ----------------- |
   * | en   | Welcome!          |
   * | ru   | Добро пожаловать! |
   */
(key: "common.welcome"): string;
}"
`;

exports[`generateTypeDefinitions > Объединяет параметры из всех языков 1`] = `
"export type TranslationKeys = "test.key"

export type TranslationParamsMap = {
  "test.key": { count: string | number | Date };
};

export interface TranslateFunctionDocs {
  /**
   * {{count}} {{total}} / {{count}}
   *
   * @translations
   * | lang | translation         |
   * | ---- | ------------------- |
   * | en   | {{count}} {{total}} |
   * | ru   | {{count}}           |
   */
(key: "test.key", params: { count: string | number | Date }): string;
}"
`;
